File: data-structures/tree/bst/bst.go
package main

import (
	"fmt"
)

type node struct {
	left  *node
	right *node
	key   int
}

func newNode(key int) *node {
	return &node{left: nil, right: nil, key: key}
}

type binarySearchTree struct {
	root *node
	size int
}

func newTree() *binarySearchTree {
	return &binarySearchTree{root: nil, size: 0}
}

func (t *binarySearchTree) insertRec(key int) *node {
	return insertRecHelper(t.root, key)
}

func insertRecHelper(n *node, key int) *node {
	if n == nil {
		n = newNode(key)
		return n
	}
	if key < n.key {
		n.left = insertRecHelper(n.left, key)
	} else {
		n.right = insertRecHelper(n.right, key)

	}
	return n
}

func (t *binarySearchTree) insert(key int) *node {
	newTreeNode := newNode(key)
	if t.root == nil {
		t.root = newTreeNode
		t.size++
		return t.root
	}
	current := t.root
	for true {
		if key < current.key {
			// go left
			if current.left == nil {
				current.left = newTreeNode
				t.size++
				return current
			}
			current = current.left
		} else {
			if current.right == nil {
				current.right = newTreeNode
				t.size++
				return current
			}
			current = current.right
		}
	}
	return current
}

func (t *binarySearchTree) getMax() int {
	if t.root == nil {
		return 0
	}
	current := t.root
	for current.right != nil {
		current = current.right
	}
	return current.key
}
func (t *binarySearchTree) getMin() int {
	if t.root == nil {
		return 0
	}
	current := t.root
	for current.left != nil {
		current = current.left
	}
	return current.key
}

func (t *binarySearchTree) bfs() []int {
	queue := []*node{}
	visited := []int{}
	var n = t.root
	queue = append(queue, n)
	for len(queue) > 0 {
		n = queue[0]
		queue = queue[1:]
		visited = append(visited, n.key)

		if n.left != nil {
			queue = append(queue, n.left)
		}
		if n.right != nil {
			queue = append(queue, n.right)
		}
	}

	return visited

}

func (t *binarySearchTree) dfsInOrder() {
	dfsInOrderRec(t.root)
}

func dfsInOrderRec(root *node) {
	if root == nil {
		return
	}
	dfsInOrderRec(root.left)
	fmt.Println(root.key)
	dfsInOrderRec(root.right)
}

func (t *binarySearchTree) dfsPreOrder() {
	dfsPreOrderRec(t.root)
}

func dfsPreOrderRec(root *node) {
	if root == nil {
		return
	}
	fmt.Println(root.key)
	dfsPreOrderRec(root.left)
	dfsPreOrderRec(root.right)
}

func (t *binarySearchTree) dfsOutOrder() {
	dfsOutOrderRec(t.root)
}

func dfsOutOrderRec(root *node) {
	if root == nil {
		return
	}
	fmt.Println(root.key)
	dfsOutOrderRec(root.left)
	dfsOutOrderRec(root.right)
}

File: data-structures/tree/bst/Bst.java

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;

class BstNode {

  public BstNode left;
  public BstNode right;
  public int value;

  public BstNode(int value) {
    this.value = value;
  }
}

public class Bst {

  BstNode root;

  public Bst() {
    this.root = null;
  }

  public void insert(int value) {
    root = inserRec(root, value);
  }

  public void printPostOrder() {
    printPostOrder(root, 0);
  }

  public void printInOrder() {
    printInOrder(root, 0);
  }

  public void printPreorder() {
    printPreorder(root, 0);
  }

  public void dfsPreOrder() {
    dfsPreOrder(root);
  }

  public ArrayList<Integer> bfs() {
    BstNode current = root;
    Queue<BstNode> queue = new LinkedList<>();
    queue.add(current);
    ArrayList<Integer> visited = new ArrayList<>();
    while (!queue.isEmpty()) {
      current = queue.remove();
      visited.add(current.value);
      if (current.left != null)
        queue.add(current.left);
      if (current.right != null)
        queue.add(current.right);
    }
    return visited;
  }

  private ArrayList<Integer> dfsPreOrder(BstNode node) {
    ArrayList<Integer> visited = new ArrayList<>();
    this.traversePreOrder(node, visited);
    System.out.println(visited);
    return visited;
  }

  private void traversePreOrder(BstNode node, ArrayList<Integer> visited) {
    visited.add(node.value);
    if (node.left != null) {
      traversePreOrder(node.left, visited);
    }
    if (node.right != null) {
      traversePreOrder(node.right, visited);
    }
  }

  private BstNode inserRec(BstNode root, int value) {
    if (root == null) {
      root = new BstNode(value);
      return root;
    }
    if (value < root.value) {
      root.left = inserRec(root.left, value);
    } else {
      root.right = inserRec(root.right, value);
    }
    return root;
  }

  private void printPostOrder(BstNode node, int level) {
    if (node == null) {
      return;
    }

    printPostOrder(node.left, level + 1);
    printPostOrder(node.right, level + 1);
    System.out.println(" ");
    System.out.println(" ".repeat(level * 2) + node.value);
  }

  private void printInOrder(BstNode node, int level) {
    if (node == null) {
      return;
    }
    printInOrder(node.left, level + 1);
    System.out.println(" ".repeat(level) + node.value + " ");
    printInOrder(node.right, level + 1);
  }

  private void printPreorder(BstNode node, int level) {
    if (node == null) {
      return;
    }
    System.out.println(" ".repeat(level) + node.value);
    printPreorder(node.left, level + 1);
    printPreorder(node.right, level + 1);
  }

  @Override
  public String toString() {
    return "Bst{" + "root=" + root + '}';
  }
}
